database/scrapers/departments.ts

```diff
@@ -0,0 +1,176 @@
+import FieldOfStudyModel from "#models/field_of_study";
+import FilesService from "#services/files_service";
+
+interface SourceResponse<T> {
+  data: T[];
+}
+
+interface DepartmentsDraft {
+  id: number;
+  name: string;
+  logo: string;
+  description: string;
+  code: string;
+  gradient_start: string;
+  gradient_end: string;
+  address: string;
+  betterCode: string;
+  createdAt: string;
+  updatedAt: string;
+}
+
+interface FieldOfStudyDraft {
+  id: number;
+  department_id: number;
+  name: string;
+  url: string;
+  isEnglish: boolean;
+  is2ndDegree: boolean;
+  isLongCycleStudies: boolean;
+  hasWeekendModeOption: boolean;
+  createdAt: string;
+  updatedAt: string;
+}
+
+interface DepartmentLinkDraft {
+  id: number;
+  department_id: number;
+  linkType: string;
+  link: string;
+  createdAt: string;
+  updatedAt: string;
+}
+
+export default class DepartmentsScraper extends BaseScraperModule {
+  static name = "Departments";
+  static description =
+    "Scrapes pwr departments, field of study and departments links data from directus";
+  static taskTitle = "Scrape departments, field of study and departments links";
+
+  private filesService = new FilesService();
+
+  private readonly directusSchemas = [
+    "Departments",
+    "FieldOfStudy",
+    "Departments_Links",
+  ];
+
+  async run(task: TaskHandle) {
+    task.update("Starting fetching all schema objects");
+
+    const [Departments, FieldOfStudy, DepartmentLink] = (await Promise.all(
+      this.directusSchemas.map((schema) =>
+        this.semaphore.runTask(() =>
+          this.fetchJSON(
+            `https://admin.topwr.solvro.pl/items/${schema}?limit=-1`,
+            schema,
+          ),
+        ),
+      ),
+    )) as [
+      SourceResponse<DepartmentsDraft>,
+      SourceResponse<FieldOfStudyDraft>,
+      SourceResponse<DepartmentLinkDraft>,
+    ];
+
+    const departmentEntries = await Promise.all(
+      Departments.data.map(async (departmentEntry) => {
+        // Logo
+        const details = (await this.fetchJSON(
+          `https://admin.topwr.solvro.pl/files/${departmentEntry.logo}?fields=filename_disk`,
+          `details for file ${departmentEntry.logo}`,
+        )) as { data: { filename_disk: string } };
+        const extension = details.data.filename_disk
+          .split(".")
+          .pop()
+          ?.toLowerCase();
+
+        const fileResponse = await this.fetchAndCheckStatus(
+          `https://admin.topwr.solvro.pl/assets/${departmentEntry.logo}`,
+          `file ${departmentEntry.logo}`,
+        );
+
+        if (fileResponse.body === null) {
+          throw new Error("Response body is null");
+        }
+        const name = await this.filesService.uploadStream(
+          Readable.fromWeb(fileResponse.body),
+          extension,
+        );
+
+        // Address
+        const regex = /\b\d{2}-\d{3}\s+[A-ZĄĆĘŁŃÓŚŹŻa-ząćęłńóśźż-]+$/;
+        const match = regex.exec(departmentEntry.address);
```

Comment:
consider moving the definition of this regex into a constant outside of this function